File: backend/src/models/session.py
from sqlalchemy import Column, Integer, String, Text, DateTime, ForeignKey, JSON
from sqlalchemy.orm import relationship
from datetime import datetime
from . import Base, TimestampMixin


class DocumentSession(Base, TimestampMixin):
    """Document session model for tracking client document progress."""
    
    __tablename__ = "document_sessions"
    
    id = Column(Integer, primary_key=True, index=True)
    client_identifier = Column(String(255), nullable=False, index=True)
    user_id = Column(Integer, ForeignKey("users.id", ondelete="CASCADE"), nullable=False)
    flow_id = Column(Integer, ForeignKey("document_flows.id", ondelete="SET NULL"), nullable=True)
    current_group_id = Column(Integer, ForeignKey("question_groups.id", ondelete="SET NULL"), nullable=True)
    is_completed = Column(Integer, default=False, nullable=False)
    completed_at = Column(DateTime, nullable=True)
    
    # Relationships
    user = relationship("User", foreign_keys=[user_id])
    flow = relationship("DocumentFlow", foreign_keys=[flow_id])
    current_group = relationship("QuestionGroup", foreign_keys=[current_group_id])
    answers = relationship("SessionAnswer", back_populates="session", cascade="all, delete-orphan")
    
    def __repr__(self) -> str:
        return f"<DocumentSession(id={self.id}, client='{self.client_identifier}', completed={self.is_completed})>"
    
    def to_dict(self) -> dict:
        """Convert session to dictionary."""
        return {
            "id": self.id,
            "client_identifier": self.client_identifier,
            "user_id": self.user_id,
            "flow_id": self.flow_id,
            "current_group_id": self.current_group_id,
            "is_completed": bool(self.is_completed),
            "completed_at": self.completed_at.isoformat() if self.completed_at else None,
            "created_at": self.created_at.isoformat(),
            "updated_at": self.updated_at.isoformat(),
        }


class SessionAnswer(Base, TimestampMixin):
    """Session answer model for storing client answers to questions."""
    
    __tablename__ = "session_answers"
    
    id = Column(Integer, primary_key=True, index=True)
    session_id = Column(Integer, ForeignKey("document_sessions.id", ondelete="CASCADE"), nullable=False, index=True)
    question_id = Column(Integer, ForeignKey("questions.id", ondelete="CASCADE"), nullable=False, index=True)
    answer_value = Column(Text, nullable=False)
    
    # Relationships
    session = relationship("DocumentSession", back_populates="answers")
    question = relationship("Question")
    
    def __repr__(self) -> str:
        return f"<SessionAnswer(id={self.id}, session_id={self.session_id}, question_id={self.question_id})>"
    
    def to_dict(self) -> dict:
        """Convert answer to dictionary."""
        return {
            "id": self.id,
            "session_id": self.session_id,
            "question_id": self.question_id,
            "answer_value": self.answer_value,
            "created_at": self.created_at.isoformat(),
            "updated_at": self.updated_at.isoformat(),
        }

File: go-backend/internal/models/session.go
package models

import (
	"time"
)

type InputForm struct {
	ID               int        `gorm:"primaryKey" json:"id"`
	ClientIdentifier string     `gorm:"size:255;index;not null" json:"client_identifier"`
	UserID           int        `gorm:"not null" json:"user_id"`
	FlowID           *int       `json:"flow_id"`
	CurrentGroupID   *int       `gorm:"column:current_group_id" json:"current_group_id"`
	IsCompleted      bool       `gorm:"default:false;not null" json:"is_completed"`
	CompletedAt      *time.Time `json:"completed_at"`
	CreatedAt        time.Time  `gorm:"not null" json:"created_at"`
	UpdatedAt        time.Time  `gorm:"not null" json:"updated_at"`

	User         User           `gorm:"foreignKey:UserID;constraint:OnDelete:CASCADE" json:"-"`
	Flow         *DocumentFlow  `gorm:"foreignKey:FlowID;constraint:OnDelete:SET NULL" json:"-"`
	CurrentGroup *QuestionGroup `gorm:"foreignKey:CurrentGroupID;constraint:OnDelete:SET NULL" json:"-"`
	Answers      []SessionAnswer `gorm:"foreignKey:SessionID;constraint:OnDelete:CASCADE" json:"answers,omitempty"`
}

func (InputForm) TableName() string { return "document_sessions" }

type InputFormResponse struct {
	ID               int        `json:"id"`
	ClientIdentifier string     `json:"client_identifier"`
	UserID           int        `json:"user_id"`
	FlowID           *int       `json:"flow_id"`
	CurrentGroupID   *int       `json:"current_group_id"`
	CurrentGroupName *string    `json:"current_group_name,omitempty"`
	IsCompleted      bool       `json:"is_completed"`
	CompletedAt      *time.Time `json:"completed_at"`
	CreatedAt        time.Time  `json:"created_at"`
	UpdatedAt        time.Time  `json:"updated_at"`
}

func (s *InputForm) ToResponse() InputFormResponse {
	return InputFormResponse{
		ID:               s.ID,
		ClientIdentifier: s.ClientIdentifier,
		UserID:           s.UserID,
		FlowID:           s.FlowID,
		CurrentGroupID:   s.CurrentGroupID,
		IsCompleted:      s.IsCompleted,
		CompletedAt:      s.CompletedAt,
		CreatedAt:        s.CreatedAt,
		UpdatedAt:        s.UpdatedAt,
	}
}

type SessionAnswer struct {
	ID          int       `gorm:"primaryKey" json:"id"`
	SessionID   int       `gorm:"index;not null" json:"session_id"`
	QuestionID  int       `gorm:"index;not null" json:"question_id"`
	AnswerValue string    `gorm:"type:text;not null" json:"answer_value"`
	CreatedAt   time.Time `gorm:"not null" json:"created_at"`
	UpdatedAt   time.Time `gorm:"not null" json:"updated_at"`

	Session  InputForm `gorm:"foreignKey:SessionID;constraint:OnDelete:CASCADE" json:"-"`
	Question Question  `gorm:"foreignKey:QuestionID;constraint:OnDelete:CASCADE" json:"-"`
}

func (SessionAnswer) TableName() string { return "session_answers" }

type AnswerSnapshot struct {
	ID             int       `gorm:"primaryKey" json:"id"`
	SessionID      int       `gorm:"index;not null" json:"session_id"`
	QuestionID     int       `gorm:"index;not null" json:"question_id"`
	QuestionNumber *string   `gorm:"type:varchar" json:"question_number"`
	AnswerValue    string    `gorm:"type:text;not null" json:"answer_value"`
	SavedAt        time.Time `gorm:"not null" json:"saved_at"`
}

func (AnswerSnapshot) TableName() string { return "answer_snapshots" }
